subsys/event_manager/event_manager.c

```diff
@@ -0,0 +1,233 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <zephyr.h>
+#include <spinlock.h>
+#include <sys/slist.h>
+#include <event_manager/event_manager.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(event_manager, CONFIG_EVENT_MANAGER_LOG_LEVEL);
+
+
+/* Event manager uses orphan sections. Below tag will allow linker to
+ * find a section usable to put the orphan sections next to.
+ */
+const struct {} linker_tag __attribute__((__section__("event_manager"))) __used;
+
+
+static void event_processor_fn(struct k_work *work);
+
+struct event_manager_event_display_bm _event_manager_event_display_bm;
+
+static K_WORK_DEFINE(event_processor, event_processor_fn);
+static sys_slist_t eventq = SYS_SLIST_STATIC_INIT(&eventq);
+static struct k_spinlock lock;
+
+
+static bool log_is_event_displayed(const struct event_type *et)
+{
+	return atomic_test_bit(_event_manager_event_display_bm.flags, et - __start_event_types);
+}
+
+static void log_event(const struct event_header *eh)
+{
+	const struct event_type *et = eh->type_id;
+
+	if (!IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENTS) ||
+	    !log_is_event_displayed(et)) {
+		return;
+	}
+
+	if (et->log_event) {
+		char log_buf[CONFIG_EVENT_MANAGER_EVENT_LOG_BUF_LEN];
+
+		int pos = et->log_event(eh, log_buf, sizeof(log_buf));
+
+		if (pos < 0) {
+			log_buf[0] = '\0';
+		} else if (pos >= sizeof(log_buf)) {
+			BUILD_ASSERT(sizeof(log_buf) >= 2,
+					 "Buffer invalid");
+			log_buf[sizeof(log_buf) - 2] = '~';
+		}
+
+		if (IS_ENABLED(CONFIG_EVENT_MANAGER_LOG_EVENT_TYPE)) {
+			LOG_INF("e: %s %s", et->name, log_strdup(log_buf));
+		} else {
+			LOG_INF("%s", log_strdup(log_buf));
+		}
+
+	} else if (IS_ENABLED(CONFIG_EVENT_MANAGER_LOG_EVENT_TYPE)) {
+		LOG_INF("e: %s", et->name);
+	}
+}
+
+static void log_event_progress(const struct event_type *et,
+			       const struct event_listener *el)
+{
+	if (!IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENTS) ||
+	    !IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENT_HANDLERS) ||
+	    !log_is_event_displayed(et)) {
+		return;
+	}
+
+	LOG_INF("|\tnotifying %s", el->name);
+}
+
+static void log_event_consumed(const struct event_type *et)
+{
+	if (!IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENTS) ||
+	    !IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENT_HANDLERS) ||
+	    !log_is_event_displayed(et)) {
+		return;
+	}
+
+	LOG_INF("|\tevent consumed");
+}
+
+static void log_event_init(void)
+{
+	if (!IS_ENABLED(CONFIG_LOG)) {
+		return;
+	}
+
+	for (const struct event_type *et = __start_event_types;
+	     (et != NULL) && (et != __stop_event_types);
+	     et++) {
+		if (et->init_log_enable) {
+			atomic_set_bit(_event_manager_event_display_bm.flags,
+					et - __start_event_types);
+		}
+	}
+}
+
+void __weak event_manager_trace_event_execution(const struct event_header *eh,
+				  bool is_start)
+{
+}
+
+void __weak event_manager_trace_event_submission(const struct event_header *eh,
+				const void *trace_info)
+{
+}
+
+void * __weak event_manager_alloc(size_t size)
```

Comment:
I think that in general we should avoid using weaks (it's a non-standard feature). Kconfig options to allow custom re-implementations would be a better choice IMO.